histogram/src/sliding_window.rs

```diff
@@ -0,0 +1,471 @@
+    /// * `slices` is the number of discrete time slices
+    ///
+    /// # Constraints:
+    /// * `n` must be less than or equal to 64
+    /// * `n` must be greater than `a + b`
+    /// * `interval` in nanoseconds must fit within a `u64`
+    /// * `interval` must be at least 1 microsecond
+    pub fn new(
+        a: u8,
+        b: u8,
+        n: u8,
+        interval: core::time::Duration,
+        slices: usize,
+    ) -> Result<Self, BuildError> {
+        let config = Config::new(a, b, n)?;
+
+        Ok(Self {
+            config,
+            interval,
+            slices,
+            started: None,
+        })
+    }
+
+    /// Specify the start time for the histogram as a `UnixInstant`.
+    pub fn start(mut self, start: UnixInstant) -> Self {
+        self.started = Some(start);
+        self
+    }
+
+    /// Consume the builder and produce a sliding window histogram that uses
+    /// atomic operations.
+    pub fn build(self) -> Result<Histogram, BuildError> {
+        let p = self.config.params();
+
+        let mut h = Histogram::new(p.a, p.b, p.n, self.interval, self.slices)?;
+
+        // if we have some start time, we move the three time fields in the
+        // histogram as necessary
+        if let Some(start) = self.started {
+            if start < h.started {
+                let delta = h.started - start;
+                h.started -= delta;
+                h.tick_origin -= delta;
+                h.tick_at.fetch_sub(delta, Ordering::Relaxed);
+            } else {
+                let delta = start - h.started;
+                h.started += delta;
+                h.tick_origin += delta;
+                h.tick_at.fetch_add(delta, Ordering::Relaxed);
+            }
+        }
+
+        Ok(h)
+    }
+}
+
+impl Histogram {
+    /// Create a new histogram that stores values across a sliding window and
+    /// allows concurrent modification.
+    ///
+    /// # Parameters:
+    /// * See the [`crate::Parameters`] documentation for the meaning of `a`,
+    ///   `b`, and `n`.
+    /// * `interval` sets the duration between snapshots. This is the minimum
+    ///   granularity in the time domain.
+    /// * `slices` sets the number of slices/snapshots that are stored. This
+    ///   controls the span of the sliding window as a multiplier for the
+    ///   `interval`.
+    ///
+    /// # Constraints:
+    /// * See the [`crate::Parameters`] documentation for the constraints for
+    ///   `a`, `b`, and `n`.
+    /// * `interval` cannot exceed 1 hour (3600 seconds)
+    /// * `interval` must be at least 1 millisecond
+    pub fn new(
+        a: u8,
+        b: u8,
+        n: u8,
+        interval: core::time::Duration,
+        slices: usize,
+    ) -> Result<Self, BuildError> {
+        let now = Instant::now();
+        let started = UnixInstant::now();
+
+        let config = Config::new(a, b, n)?;
+
+        let mut live = Vec::with_capacity(config.total_bins());
```

Comment:
Should we move this to after the interval checks? I think lifetimes takes care of it so there isn't a memory leak even if it does return, but seems cleaner that way.